File: backend/app/services/scoring/attractiveness.py
import hashlib

from functools import lru_cache


# Inputs the weighted formula below consumes, in formula order.
SCORING_INPUTS = (
    "attention_duration",
    "interaction_frequency",
    "pickup_rate",
    "conversion_rate",
    "repeat_engagement",
)

# The four inputs the prototype generator derives from attention duration.
GENERATED_METRICS = (
    "interaction_frequency",
    "repeat_engagement",
    "pickup_rate",
    "conversion_rate",
)

# Reference range that maps average dwell time onto the 0-100 scale the
# formula expects. A calibration bound for a metric the pipeline already
# records — no new metric is introduced here. Tune it if the store layout or
# camera framing changes materially.
#
# The upper bound is the dwell that counts as a fully engaged shelf visit,
# not the longest dwell a camera can record. It was 60s, which is loiter
# time rather than shopping time: a shopper studying a bay for ten seconds
# scored 17 on the attention axis, every downstream input inherited that,
# and every shelf in the estate came out in the Critical band regardless of
# how it was actually performing. Twelve seconds is the engaged-visit mark
# the frontend model uses for the same purpose (DWELL_REFERENCE_SECONDS in
# src/lib/attractiveness.ts), so both ends of the system now agree on what
# a fully engaged visit looks like.
#
# The weighted formula below is untouched by this: it still consumes five
# 0-100 inputs and weights them exactly as specified. Only the calibration
# that turns seconds into the attention input has changed.
DWELL_TIME_REFERENCE = (0.0, 12.0)      # seconds per session

# Identifier used when scoring across every zone rather than one shelf, so the
# store-wide figure is still stable and distinct from any single shelf.
ALL_ZONES_IDENTIFIER = "__all_zones__"

# Per-metric shape of the generated values:
#   slope  - how strongly the metric tracks attention duration
#   spread - total width of the deterministic variation around that trend
# Slopes descend deliberately: a shopper looking at a shelf interacts with it
# more often than they pick something up, and picks up more often than they
# buy. Spread widens further down the funnel for the same reason.
METRIC_PROFILE = {
    "interaction_frequency": (0.85, 18.0),
    "repeat_engagement": (0.70, 24.0),
    "pickup_rate": (0.60, 28.0),
    "conversion_rate": (0.45, 30.0),
}


def normalize(value, min_value, max_value):
    """
    Convert a metric into a 0-100 range.
    """

    if max_value == min_value:
        return 50.0

    score = (
        (value - min_value)
        / (max_value - min_value)
    ) * 100

    return max(0.0, min(100.0, score))


def calculate_attractiveness_score(
    attention_duration,
    interaction_frequency,
    pickup_rate,
    conversion_rate,
    repeat_engagement,
):
    """
    Calculate product attractiveness score.

    All inputs should already be normalized
    between 0 and 100.
    """

    score = (
        attention_duration * 0.35
        + interaction_frequency * 0.25
        + pickup_rate * 0.20
        + conversion_rate * 0.15
        + repeat_engagement * 0.05
    )

    return round(
        max(0.0, min(100.0, score)),
        2
    )


def _unit_interval(identifier, attention_duration, metric):
    """
    A stable value in [0, 1) derived from the inputs.

    Hashing rather than random(): the same shelf and the same attention
    duration always produce the same number, while a different shelf or a
    different attention duration produces an unrelated one. Nothing here reads
    a clock, a seed or process state.
    """

    key = f"{identifier}|{attention_duration:.4f}|{metric}".encode("utf-8")

    digest = hashlib.blake2b(key, digest_size=8).digest()

    return int.from_bytes(digest, "big") / float(1 << 64)


@lru_cache(maxsize=512)
def _generate_metrics(identifier, attention_duration):
    """
    Cached generation keyed on (identifier, attention duration).

    Because those two are the only inputs, the work is redone only when a new
    video changes the recorded attention duration for a shelf — repeated
    scoring calls for unchanged analytics are served from the cache.

    Returns a tuple of pairs so the cached value cannot be mutated by callers.
    """

    generated = []

    for metric in GENERATED_METRICS:

        slope, spread = METRIC_PROFILE[metric]

        # Centred on the attention-driven trend, then nudged by a deterministic
        # offset in [-spread/2, +spread/2]. The trend dominates, so the
        # correlation with attention duration is real rather than a fixed
        # offset, while no two shelves land on exactly the same curve.
        offset = (
            _unit_interval(identifier, attention_duration, metric) - 0.5
        ) * spread

        value = attention_duration * slope + offset

        generated.append(
            (metric, round(max(0.0, min(100.0, value)), 2))
        )

    return tuple(generated)


def generate_prototype_metrics(identifier, attention_duration):
    """
    Derive the four non-observable scoring inputs from attention duration.

    The vision pipeline cannot see a product being handled or bought, and this
    prototype has no POS feed, so interaction frequency, repeat engagement,
    pickup rate and conversion rate are generated from the one real signal
    there is. Values are deterministic, correlated with attention duration and
    bounded to 0-100.

    identifier is a stable shelf key (zone value or name); attention_duration
    is the already-normalized 0-100 figure.
    """

    return dict(
        _generate_metrics(
            identifier or ALL_ZONES_IDENTIFIER,
            round(float(attention_duration), 4),
        )
    )


def resolve_scoring_inputs(engagement, identifier=None):
    """
    Build the full set of formula inputs for a shelf zone.

    Attention duration is the one input taken from real analytics; the other
    four are generated from it. Returns (metrics, sources) where metrics feeds
    calculate_attractiveness_score unchanged and sources labels each input
    "analytics", "generated" or "unavailable" for the UI.
    """

    attention = 0.0

    if engagement:
        attention = round(
            normalize(
                engagement["avg_dwell_time"],
                *DWELL_TIME_REFERENCE
            ),
            2
        )

    metrics = {"attention_duration": attention}

    sources = {
        "attention_duration": "analytics" if engagement else "unavailable"
    }

    generated = generate_prototype_metrics(identifier, attention)

    for metric in GENERATED_METRICS:
        metrics[metric] = generated[metric]
        sources[metric] = "generated"

    # Keep formula order so the returned mapping reads the same as before.
    return (
        {field: metrics[field] for field in SCORING_INPUTS},
        {field: sources[field] for field in SCORING_INPUTS},
    )

File: frontend/src/lib/attractiveness.ts
import type { MetricSource, ScoringMetrics } from "@/lib/api";
import type { Tone } from "@/lib/tone";

/* ═══════════════════════════════════════════════════════════════════════
   Product attractiveness — scoring, health banding and advice.

   One model, consumed by every surface in the app: the attractiveness
   cards, the recommendation panel, the KPI row, the report zones and the
   system health alerts. Because they all read the figures produced here,
   no two panels can state a different score or a different severity for
   the same shelf.

   ── The formula ──────────────────────────────────────────────────────
   Exactly as specified, and identical to
   calculate_attractiveness_score() in
   backend/app/services/scoring/attractiveness.py:

     score = attention_duration    * 0.35
           + interaction_frequency * 0.25
           + pickup_rate           * 0.20
           + conversion_rate       * 0.15
           + repeat_engagement     * 0.05

   The score is NEVER adjusted, rescaled or remapped after this sum. The
   only thing this module ever substitutes is a missing or unusable
   *input*; the formula then runs on those inputs untouched.

   ── Why inputs are sometimes substituted ─────────────────────────────
   The vision pipeline can only observe attention. Interaction, pickup,
   conversion and repeat engagement need a POS feed the prototype does not
   have, so the backend derives them from attention duration.

   That derivation is now usable on its own: the backend normalizes dwell
   against a 12-second engaged visit, so a real shelf lands across the whole
   0-100 range rather than bunching into single digits. Measured shelves are
   therefore scored from the backend's own inputs, untouched — which is what
   lets the notification bell and the exported reports state the same score
   as this screen.

   Substitution is now the exception it was meant to be: a shelf with no
   metrics at all, or with too few sessions to be representative, still gets
   a realistic set (deterministically, anchored to whatever real signal
   exists). Those are labelled "generated", which the attractiveness cards
   surface with an "est" marker — so an estimated shelf is visibly distinct
   from a measured one.
   ═══════════════════════════════════════════════════════════════════════ */

export const METRIC_KEYS = [
  "attention_duration",
  "interaction_frequency",
  "pickup_rate",
  "conversion_rate",
  "repeat_engagement",
] as const;

export type MetricKey = (typeof METRIC_KEYS)[number];

/** Formula weights. These are the specification; do not tune them. */
export const SCORING_WEIGHTS: Record<MetricKey, number> = {
  attention_duration: 0.35,
  interaction_frequency: 0.25,
  pickup_rate: 0.2,
  conversion_rate: 0.15,
  repeat_engagement: 0.05,
};

/** Display names for the five inputs, used in advice and alert copy. */
export const METRIC_LABELS: Record<MetricKey, string> = {
  attention_duration: "Attention",
  interaction_frequency: "Interaction",
  pickup_rate: "Pickup",
  conversion_rate: "Conversion",
  repeat_engagement: "Repeat",
};

/**
 * The weighted formula. Clamped to 0-100 and rounded to two decimals, so a
 * frontend-computed score is bit-comparable with the backend's.
 */
export function attractivenessScore(metrics: ScoringMetrics): number {
  const total = METRIC_KEYS.reduce(
    (sum, key) => sum + (metrics[key] ?? 0) * SCORING_WEIGHTS[key],
    0
  );

  return Math.round(Math.max(0, Math.min(100, total)) * 100) / 100;
}

/* ── Health bands ───────────────────────────────────────────────────────
   Derived from the final computed score and nothing else, which is what
   guarantees severity can never contradict the number beside it. A shelf
   scoring 90 cannot raise a High-priority alert.
   ─────────────────────────────────────────────────────────────────────── */

export type HealthBand = {
  /** Status shown on the score card. */
  label: string;
  /** Priority shown on the advice card. */
  priority: "Low" | "Medium" | "High" | "Critical";
  tone: Tone;
  /** True when the band calls for corrective action rather than upkeep. */
  corrective: boolean;
  /** How many weak inputs the advice should address. */
  focusCount: number;
};

const BANDS: { min: number; band: HealthBand }[] = [
  {
    min: 85,
    band: {
      label: "Healthy",
      priority: "Low",
      tone: "healthy",
      corrective: false,
      focusCount: 0,
    },
  },
  {
    min: 70,
    band: {
      label: "Good",
      priority: "Medium",
      tone: "warning",
      corrective: true,
      focusCount: 1,
    },
  },
  {
    min: 55,
    band: {
      label: "Monitor",
      priority: "Medium",
      tone: "warning",
      corrective: true,
      focusCount: 2,
    },
  },
  {
    min: 35,
    band: {
      label: "Needs Attention",
      priority: "High",
      tone: "critical",
      corrective: true,
      focusCount: 3,
    },
  },
  {
    min: 0,
    band: {
      label: "Critical",
      priority: "Critical",
      tone: "critical",
      corrective: true,
      focusCount: 3,
    },
  },
];

/** Band for a computed score. Single source of truth for severity. */
export function healthFor(score: number): HealthBand {
  return (BANDS.find((entry) => score >= entry.min) ?? BANDS[BANDS.length - 1])
    .band;
}

/* ── Input resolution ───────────────────────────────────────────────────
   Substitution happens here and only here.
   ─────────────────────────────────────────────────────────────────────── */

/** Fewer sessions than this and the inputs are not yet representative. */
const MIN_SESSIONS = 5;

/** Dwell, in seconds, that reads as a fully engaged shelf visit. */
const DWELL_REFERENCE_SECONDS = 12;

/** Session count that reads as a well-sampled shelf. */
const SESSION_REFERENCE = 40;

/**
 * Realistic retail range for each input, and how widely shelves vary on it.
 *
 * Ranges narrow down the funnel: a shopper looks at a shelf more often than
 * they touch it, touches more often than they buy, and buys more often than
 * they come back. Spread widens for the same reason — the further down the
 * funnel, the more one shelf differs from the next.
 */
const INPUT_RANGES: Record<
  MetricKey,
  { min: number; max: number; spread: number }
> = {
  attention_duration: { min: 30, max: 94, spread: 8 },
  interaction_frequency: { min: 24, max: 90, spread: 10 },
  pickup_rate: { min: 20, max: 86, spread: 11 },
  conversion_rate: { min: 16, max: 82, spread: 12 },
  repeat_engagement: { min: 12, max: 78, spread: 13 },
};

export type ShelfInput = {
  /** Stable identity — the shelf label or zone value. */
  key: string;
  /** `metrics_used` from the backend, when the response carries it. */
  metrics?: ScoringMetrics | null;
  /** Sessions the backend scored from. */
  sessions?: number | null;
  /** Average dwell in seconds, when the caller has it. */
  dwellSeconds?: number | null;
};

function clamp01(value: number): number {
  return Number.isFinite(value) ? Math.min(1, Math.max(0, value)) : 0;
}

/**
 * A stable value in [0, 1) derived from a string.
 *
 * FNV-1a, mirroring the intent of the backend's blake2b helper: the same
 * shelf always produces the same number, a different shelf produces an
 * unrelated one, and nothing reads a clock, a seed or process state. This
 * is what makes generated inputs identical on every refresh.
 */
function hashUnit(seed: string): number {
  let hash = 0x811c9dc5;

  for (let index = 0; index < seed.length; index += 1) {
    hash ^= seed.charCodeAt(index);
    hash = Math.imul(hash, 0x01000193) >>> 0;
  }

  return hash / 0x100000000;
}

/**
 * How well this shelf is performing, in [0, 1], from whatever the pipeline
 * actually measured.
 *
 * Dwell and session volume are real signals, so they lead. The hash only
 * fills the remainder, and takes over entirely for a shelf with no
 * measurements at all — so two unmeasured shelves still differ from each
 * other, stably.
 */
function performanceAnchor(input: ShelfInput): number {
  const sessions = input.sessions;

  // attention_duration is normalize(avg_dwell, 0, 60), which inverts. It is
  // preferred over a caller-supplied dwell precisely because every caller has
  // it: the scoring response and the report response both carry it, rounded
  // the same way, so two panels scoring the same shelf start from the same
  // number even though only one of them was handed a dwell figure. Taking the
  // caller's dwell first would let those two panels disagree in the decimals.
  const attention = input.metrics?.attention_duration;

  const dwellSeconds =
    typeof attention === "number"
      ? (attention / 100) * 60
      : // No metrics at all — a caller that measured dwell itself can still
        // anchor the shelf with it.
        (input.dwellSeconds ?? null);

  const baseline = 0.3 + hashUnit(`${input.key}|anchor`) * 0.5;

  let anchor = 0;
  let weight = 0;

  if (dwellSeconds != null) {
    anchor += clamp01(dwellSeconds / DWELL_REFERENCE_SECONDS) * 0.5;
    weight += 0.5;
  }

  if (sessions != null && sessions > 0) {
    anchor +=
      clamp01(Math.log1p(sessions) / Math.log1p(SESSION_REFERENCE)) * 0.3;
    weight += 0.3;
  }

  // Whatever is left over is carried by the shelf's own stable baseline.
  return clamp01(anchor + baseline * (1 - weight));
}

/**
 * True when the backend's inputs cannot carry a meaningful score.
 *
 * "Sparse" means missing or unrepresentative — no metrics at all, or too few
 * sessions behind them. It deliberately does not mean "low": a shelf the
 * cameras genuinely watched shoppers walk past without stopping scores
 * badly, and that is a finding, not a gap in the data. Substituting there
 * would replace a real measurement with an invented one and would put a
 * comfortable number over a shelf that needs attention.
 *
 * A score threshold used to sit here as well, because the backend
 * normalized dwell against a 60-second reference and every real shelf came
 * out in single digits. That reference is now calibrated to a 12-second
 * engaged visit (DWELL_TIME_REFERENCE in the backend scoring service), so
 * measured shelves land across the usable range and no longer need
 * rescuing.
 */
export function inputsAreSparse(input: ShelfInput): boolean {
  const metrics = input.metrics;

  if (!metrics) return true;

  if (METRIC_KEYS.some((key) => typeof metrics[key] !== "number")) return true;

  return (input.sessions ?? 0) < MIN_SESSIONS;
}

/**
 * The five formula inputs for a shelf, and where each came from.
 *
 * Real backend metrics are returned untouched whenever they are usable.
 * Otherwise a realistic set is generated, anchored to the shelf's measured
 * performance so the ordering still follows the analytics.
 */
export function resolveInputs(input: ShelfInput): {
  metrics: ScoringMetrics;
  sources: Record<MetricKey, MetricSource>;
  generated: boolean;
} {
  if (!inputsAreSparse(input) && input.metrics) {
    const metrics = input.metrics;

    return {
      metrics,
      sources: METRIC_KEYS.reduce(
        (all, key) => ({ ...all, [key]: "analytics" as MetricSource }),
        {} as Record<MetricKey, MetricSource>
      ),
      generated: false,
    };
  }

  const anchor = performanceAnchor(input);

  const metrics = METRIC_KEYS.reduce((all, key) => {
    const range = INPUT_RANGES[key];

    // Centred on the shelf's performance, then nudged by a deterministic
    // offset so no two shelves sit on exactly the same curve.
    const offset = (hashUnit(`${input.key}|${key}`) - 0.5) * range.spread;

    const value = range.min + (range.max - range.min) * anchor + offset;

    return {
      ...all,
      [key]:
        Math.round(
          Math.max(range.min, Math.min(range.max, value)) * 100
        ) / 100,
    };
  }, {} as ScoringMetrics);

  return {
    metrics,
    sources: METRIC_KEYS.reduce(
      (all, key) => ({ ...all, [key]: "generated" as MetricSource }),
      {} as Record<MetricKey, MetricSource>
    ),
    generated: true,
  };
}

/* ── Advice ─────────────────────────────────────────────────────────────
   Driven by the weakest contributing inputs and the final score. Never
   random, never a fixed list.
   ─────────────────────────────────────────────────────────────────────── */

const PLAYBOOK: Record<MetricKey, string[]> = {
  attention_duration: [
    "Raise visibility: move the range to eye level so shoppers register it on approach.",
    "Add shelf-edge signage and a clearer category header to draw the eye.",
    "Check lighting on this bay — poorly lit fixtures suppress dwell time.",
  ],
  interaction_frequency: [
    "Review packaging and facings so the range reads clearly at a distance.",
    "Add a promotional display or end-cap tie-in to invite engagement.",
    "Trial a demo or sampling touchpoint to lift hands-on interaction.",
  ],
  pickup_rate: [
    "Improve accessibility: bring stock forward and clear obstructions at reach height.",
    "Reorganise the shelf by category so individual products are easy to isolate.",
    "Reduce facing density — crowded bays measurably suppress pick-up.",
  ],
  conversion_rate: [
    "Review pricing against the alternatives sitting beside it.",
    "Add a clear offer or multi-buy to close the decision at the shelf.",
    "Expand shelf-edge product information so shoppers can compare without help.",
  ],
  repeat_engagement: [
    "Broaden the assortment so returning shoppers find something new.",
    "Tie this bay into the loyalty programme with a targeted reward.",
    "Improve the post-purchase experience to bring this category's shoppers back.",
  ],
};

/** Upkeep advice for a shelf that is already performing. */
const MAINTENANCE = [
  "Performance is strong — hold the current planogram.",
  "Keep facings stocked and review weekly for drift.",
  "Use this bay as the reference layout for comparable fixtures.",
];

export type WeakInput = {
  metric: MetricKey;
  label: string;
  value: number;
  /** Points of score lost on this input: weight × (100 − value). */
  shortfall: number;
};

/**
 * The inputs costing this shelf the most score, worst first.
 *
 * Ranked by weighted shortfall rather than raw value, so "weakest" means
 * "holding the score down hardest" — a mediocre attention figure outranks a
 * poor repeat figure, because attention carries seven times the weight.
 */
export function weakestInputs(metrics: ScoringMetrics): WeakInput[] {
  return METRIC_KEYS.map((metric) => {
    const value = metrics[metric] ?? 0;

    return {
      metric,
      label: METRIC_LABELS[metric],
      value,
      shortfall:
        Math.round(SCORING_WEIGHTS[metric] * (100 - value) * 100) / 100,
    };
  }).sort((a, b) => b.shortfall - a.shortfall);
}

/** "a", "a and b", "a, b and c" — for naming the inputs being addressed. */
function listOf(items: string[]): string {
  if (items.length <= 1) return items[0] ?? "";

  return `${items.slice(0, -1).join(", ")} and ${items[items.length - 1]}`;
}

/**
 * Advice for one shelf.
 *
 * A healthy shelf gets maintenance and monitoring; every other band gets
 * one corrective action per weak input, as many as the band calls for. The
 * action chosen from each playbook is picked by a stable hash of the shelf
 * and the metric, so it never shuffles between refreshes but does differ
 * between shelves.
 */
export function adviceFor(
  key: string,
  metrics: ScoringMetrics,
  score: number
): string[] {
  const band = healthFor(score);

  if (!band.corrective) {
    return MAINTENANCE;
  }

  const focus = weakestInputs(metrics).slice(0, band.focusCount);

  const actions = focus.map((weak) => {
    const options = PLAYBOOK[weak.metric];

    return options[Math.floor(hashUnit(`${key}|${weak.metric}|advice`) * options.length) % options.length];
  });

  // Always close with the monitoring line, so even a corrective card says
  // what to do after the fix.
  actions.push(
    `Re-measure ${listOf(
      focus.map((weak) => weak.label.toLowerCase())
    )} after the next footage run to confirm the change landed.`
  );

  return actions;
}

/* ── The assessment every surface consumes ──────────────────────────── */

export type ShelfAssessment = {
  key: string;
  metrics: ScoringMetrics;
  sources: Record<MetricKey, MetricSource>;
  /** True when the five inputs were substituted. */
  generated: boolean;
  /** Straight from the weighted formula. Never adjusted afterwards. */
  score: number;
  health: HealthBand;
  weakest: WeakInput[];
  recommendations: string[];
  /** Present only for bands that call for escalation. */
  alert: { severity: string; message: string } | null;
};

export function assessShelf(input: ShelfInput): ShelfAssessment {
  const { metrics, sources, generated } = resolveInputs(input);

  // The formula, run on the resolved inputs. Nothing touches the result.
  const score = attractivenessScore(metrics);

  const health = healthFor(score);
  const weakest = weakestInputs(metrics);
  const recommendations = adviceFor(input.key, metrics, score);

  // Alerts escalate only from the two lowest bands, so a shelf scoring in
  // the Good or Healthy range can never appear in the alert list.
  const alerting = health.priority === "High" || health.priority === "Critical";

  return {
    key: input.key,
    metrics,
    sources,
    generated,
    score,
    health,
    weakest,
    recommendations,
    alert: alerting
      ? {
          severity: health.priority,
          message: `${health.label}: ${weakest[0].label.toLowerCase()} is the largest drag on this shelf, costing ${weakest[0].shortfall.toFixed(1)} points.`,
        }
      : null,
  };
}

/** Assess a set of shelves, keyed for lookup. */
export function assessShelves(
  inputs: ShelfInput[]
): Map<string, ShelfAssessment> {
  return new Map(inputs.map((input) => [input.key, assessShelf(input)]));
}
